mmhuman3d/models/registrants/smplifyx.py

```diff
@@ -0,0 +1,473 @@
+            keypoints3d_weight: weight of 3D keypoint loss
+            shape_prior_weight: weight of shape prior loss
+            joint_prior_weight: weight of joint prior loss
+            smooth_loss_weight: weight of smooth loss
+            pose_prior_weight: weight of pose prior loss
+            joint_weights: per joint weight of shape (K, )
+            return_verts: whether to return vertices
+            return_joints: whether to return joints
+            return_full_pose: whether to return full pose
+            reduction_override: reduction method, e.g., 'none', 'sum', 'mean'
+
+        Returns:
+            ret: a dictionary that includes body model parameters,
+                and optional attributes such as vertices and joints
+        """
+
+        ret = {}
+
+        body_model_output = self.body_model(
+            global_orient=global_orient,
+            body_pose=body_pose,
+            betas=betas,
+            transl=transl,
+            left_hand_pose=left_hand_pose,
+            right_hand_pose=right_hand_pose,
+            expression=expression,
+            jaw_pose=jaw_pose,
+            leye_pose=leye_pose,
+            reye_pose=reye_pose,
+            return_verts=return_verts,
+            return_full_pose=return_full_pose)
+
+        model_joints = body_model_output['joints']
+        model_joint_mask = body_model_output['joint_mask']
+
+        loss_dict = self._compute_loss(
+            model_joints,
+            model_joint_mask,
+            keypoints2d=keypoints2d,
+            keypoints2d_conf=keypoints2d_conf,
+            keypoints2d_weight=keypoints2d_weight,
+            keypoints3d=keypoints3d,
+            keypoints3d_conf=keypoints3d_conf,
+            keypoints3d_weight=keypoints3d_weight,
+            joint_prior_weight=joint_prior_weight,
+            shape_prior_weight=shape_prior_weight,
+            smooth_loss_weight=smooth_loss_weight,
+            pose_prior_weight=pose_prior_weight,
+            joint_weights=joint_weights,
+            reduction_override=reduction_override,
+            body_pose=body_pose,
+            betas=betas)
+        ret.update(loss_dict)
+
+        if return_verts:
+            ret['vertices'] = body_model_output['vertices']
+        if return_full_pose:
+            ret['full_pose'] = body_model_output['full_pose']
+        if return_joints:
+            ret['joints'] = model_joints
+
+        return ret
+
+    def _set_keypoint_idxs(self):
+        """Set keypoint indices to 1) body parts to be assigned different
+        weights 2) be ignored for keypoint loss computation.
+
+        Returns:
+            None
+        """
+        convention = self.body_model.keypoint_dst
+
+        # obtain ignore keypoint indices
+        if self.ignore_keypoints is not None:
+            self.ignore_keypoint_idxs = []
+            for keypoint_name in self.ignore_keypoints:
+                keypoint_idx = get_keypoint_idx(
```

Comment:
Assume `len(self.ignore_keypoints)==m` and `len(convention)==n`,  the current method consumes time O(m*n).
As self.ignore_keypoints is a dict using hash table, we could use code below to reduce time to O(1*n):
`
for keypoint_idx, keypoint_name in enumerate(convention_names_list):   
    if keypoint_name in self.ignore_keypoints:   
      self.ignore_keypoint_idxs.append(keypoint_idx)   
`

Reply:
Refer to issue #64 .